library/core/src/array/mod.rs

```diff
@@ -504,7 +581,14 @@ where
     let mut guard: Guard<_, N> =
         Guard { ptr: MaybeUninit::slice_as_mut_ptr(&mut array), initialized: 0 };
```

Comment:
AFAIK, this pointer from `&mut array` will be invalidated by any subsequent mutable access to `array` that doesn’t go through that pointer. I.e. the `array.get_unchecked_mut`. Miri seems to agree that there’s a problem:

<details><summary>Details to reproduce miri error (click to expand)</summary>

If you put
```rs
#![feature(
    maybe_uninit_uninit_array,
    maybe_uninit_slice,
    maybe_uninit_array_assume_init
)]

fn main() {
    let mut flag = false;
    let x: Option<Result<[Box<u8>; 2], ()>> = collect_into_array(&mut iter::from_fn(|| {
        Some(if !flag {
            flag = true;
            Ok(Box::new(42))
        } else {
            Err(())
        })
    }));
}

use std::{
    iter,
    mem::{self, MaybeUninit},
};

/// Pulls `N` items from `iter` and returns them as an array. If the iterator
/// yields fewer than `N` items, `None` is returned and all already yielded
/// items are dropped.
///
/// Since the iterator is passed as a mutable reference and this function calls
/// `next` at most `N` times, the iterator can still be used afterwards to
/// retrieve the remaining items.
///
/// If `iter.next()` panicks, all items already yielded by the iterator are
/// dropped.
fn collect_into_array<E, I, T, const N: usize>(iter: &mut I) -> Option<Result<[T; N], E>>
where
    I: Iterator<Item = Result<T, E>>,
{
    if N == 0 {
        // SAFETY: An empty array is always inhabited and has no validity invariants.
        return unsafe { Some(Ok(mem::zeroed())) };
    }

    struct Guard<T, const N: usize> {
        ptr: *mut T,
        initialized: usize,
    }

    impl<T, const N: usize> Drop for Guard<T, N> {
        fn drop(&mut self) {
            debug_assert!(self.initialized <= N);

            let initialized_part = std::ptr::slice_from_raw_parts_mut(self.ptr, self.initialized);

            // SAFETY: this raw slice will contain only initialized objects.
            unsafe {
                std::ptr::drop_in_place(initialized_part);
            }
        }
    }

    let mut array = MaybeUninit::uninit_array::<N>();
    let mut guard: Guard<_, N> = Guard {
        ptr: MaybeUninit::slice_as_mut_ptr(&mut array),
        initialized: 0,
    };

    while let Some(item_rslt) = iter.next() {
        let item = match item_rslt {
            Err(err) => {
                return Some(Err(err));
            }
            Ok(elem) => elem,
        };

        // SAFETY: `guard.initialized` starts at 0, is increased by one in the
        // loop and the loop is aborted once it reaches N (which is
        // `array.len()`).
        unsafe {
            array.get_unchecked_mut(guard.initialized).write(item);
        }
        guard.initialized += 1;

        // Check if the whole array was initialized.
        if guard.initialized == N {
            mem::forget(guard);

            // SAFETY: the condition above asserts that all elements are
            // initialized.
            let out = unsafe { MaybeUninit::array_assume_init(array) };
            return Some(Ok(out));
        }
    }

    // This is only reached if the iterator is exhausted before
    // `guard.initialized` reaches `N`. Also note that `guard` is dropped here,
    // dropping all already initialized elements.
    None
}
```
into a crate and run `MIRIFLAGS="-Zmiri-track-raw-pointers" cargo miri run`, you’ll get something like
```
error: Undefined Behavior: trying to reborrow for SharedReadWrite at alloc1375, but parent tag <2950> does not have an appropriate item in the borrow stack
   --> /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:188:1
    |
188 | pub unsafe fn drop_in_place<T: ?Sized>(to_drop: *mut T) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for SharedReadWrite at alloc1375, but parent tag <2950> does not have an appropriate item in the borrow stack
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
            
    = note: inside `std::ptr::drop_in_place::<[std::boxed::Box<u8>]> - shim(Some([std::boxed::Box<u8>]))` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:188:1
note: inside `<collect_into_array::Guard<std::boxed::Box<u8>, 2_usize> as std::ops::Drop>::drop` at src/main.rs:56:17
   --> src/main.rs:56:17
    |
56  |                 std::ptr::drop_in_place(initialized_part);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: inside `std::ptr::drop_in_place::<collect_into_array::Guard<std::boxed::Box<u8>, 2_usize>> - shim(Some(collect_into_array::Guard<std::boxed::Box<u8>, 2_usize>))` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:188:1
note: inside `collect_into_array::<(), std::iter::FromFn<[closure@src/main.rs:9:85: 16:6]>, std::boxed::Box<u8>, 2_usize>` at src/main.rs:98:1
   --> src/main.rs:98:1
    |
98  | }
    | ^
note: inside `main` at src/main.rs:9:47
   --> src/main.rs:9:47
    |
9   |       let x: Option<Result<[Box<u8>; 2], ()>> = collect_into_array(&mut iter::from_fn(|| {
    |  _______________________________________________^
10  | |         Some(if !flag {
11  | |             flag = true;
12  | |             Ok(Box::new(42))
...   |
15  | |         })
16  | |     }));
    | |_______^
    = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
    = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
    = note: inside closure at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:64:18
    = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:259:13
    = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:403:40
    = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:367:19
    = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:133:14
    = note: inside closure at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:46:48
    = note: inside `std::panicking::r#try::do_call::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:403:40
    = note: inside `std::panicking::r#try::<isize, [closure@std::rt::lang_start_internal::{closure#2}]>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:367:19
    = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#2}], isize>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:133:14
    = note: inside `std::rt::lang_start_internal` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:46:20
    = note: inside `std::rt::lang_start::<()>` at /home/frank/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:63:17
```

</details>

A better option might be to put a mutable reference (`&'a mut [MaybeUninit<T>; N]`) into the `Guard` and only access the array through that guard. This has the additional advantage of avoiding unnecessary use of raw pointers.

Reply:
👍

I changed the field from raw mutable pointer to mutable reference but such change may require a dedicated PR

Reply:
@c410-f3r ha, you were quicker, I’ve made a commit myself, too: cdf54871f452c78845db0ca9d91cf16c09e43b57

Still, consider changing your fix to that: No need to use raw pointers when there’s `MaybeUninit::slice_assume_init_mut`; your fix doesn’t do anything in the drop_in_place call AFAICT, because it’s creating a slice of maybeuninits. The `: Guard<_, _, N>` type annotation is redundant.

Reply:
I.e. these changes could still be done: https://github.com/c410-f3r/rust/pull/2

Reply:
Thanks for suggesting these changes. If desired, feel to create a separated PR since this particular set of code is currently being used on production.

Reply:
> If desired, feel to create a separated PR 

I will do so in case this PR starts to stagnate again